docs/rfcs/700-config-features-and-defaultpairs-via-connection-string.md

```diff
@@ -0,0 +1,225 @@
+The connection string exposes basic types and not support complex types like `struct` at present. So we can't set `ServiceFeature`/`StorageFeatures` and `DefaultServicePairs`/`DefaultStoragePairs` by connection string during initialization.
+
+`Default*Pairs` is a slightly complex struct:
+
+```go
+// DefaultStoragePairs is default pairs for specific action
+type DefaultStoragePairs struct {
+	Create            []Pair
+	Delete            []Pair
+	List              []Pair
+	Metadata          []Pair
+	Read              []Pair
+	Stat              []Pair
+	Write             []Pair
+}
+```
+
+A possible format that follows the struct in connection string for default storage paris may be:
+
+ `default_storage_pairs={write:[storage_class=STANDARD_IA,key2=value],read:[]}`
+
+Obviously, It's complex for users and hard to pass in a valid string. In addition, it either exposes the internal details to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers).
+
+## Proposal
+
+I propose to support config features and default pairs in plain string format in connection string.
+
+### Features
+
+The format of features pairs in connection string is:
+
+`enable_key=value`
+
+- The whole key of the pair SHOULD have the prefix `enable_`.
+- `key` is the pair name defined in `features.toml` and supported in service, and the format SHOULD be exactly the same.
+- `value` SHOULD be `true` or `false`.
+
+So a valid connection string containing features could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&enable_loose_pair=false&enable_virtual_dir=true`
+
+The connection string above is equivalent to:
+
+```go
+store, err := s3.NewStorage(
+	ps.WithCredential("hmac:xxxx:xxxx"),
+	ps.WithName("bucket_name"),
+	ps.WithWorkDir("/prefix"),
+	s3.WithStorageFeatures(s3.StorageFeaturs{LoosePair:  true,
+		LoosePair:  true,
+		VirtualDir: true,
+    })
+)
+```
+
+### DefaultPairs
+
+The format of default pair in connection string is:
+
+`default_key=value`
+
+- The whole key of the pair SHOULD have the prefix `default_`.
+- `key` is the pair name defined in `toml` and the format SHOULD be exactly the same.
+- Operation has a different set of default pairs. Pairs in operation pair list will share the same default value if have the same name .
+
+So a valid connection string containing default pairs could be:
+
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&default_server_side_encryption=xxxx&default_strorage_class=xxxx`
+
+Also, we generate the following function for default pairs to support configure the default pair during initialization in the way of `WithXxx()`:
+
+Take `default_storage_class` for example:
+
+```go
+func WithDefaultStorageClass(v string) Pair {
+	reutrn Pair{
+		Key:   "default_storage_class",
+		Value: v,
+	}
+}
+```

```

Comment:
What's its relationship with the old `DefaultStoragePairs`?

Reply:
I think `DefaultStoragePairs` is more fine-grained compaired to `WithDefaultStorageClass`.
`DefaultStoragePairs` is an embeded struct contains a set of `[]Pair` for supported operations. So users need to assign value to pairs of each operation one by one, even when different operations have the same pair.
For `WithDefaultStorageClass`,  pairs that have the same name will share the same default value，users needn't to specify the operation.

Reply:
The two kind of pairs can be used together (not orthogonal). Will it make users confused? Will they conflict with each other?

Reply:
There's no confilic. I found that there's no need to generate `WithDefaultXxx()` functions. Maybe we can just use the existing functions `WithXxx()` to set the default paris?

Reply:
> Will they conflict with each other?

When both `WithDefault*Pairs()` and `WithXxx()` are used, I think the values passed in by `WithDefault*Pairs()` should be picked (the priority is: `pairs from args` > `WithDefault*Pairs()` > `WithXxx`), and this can be and should be generated.

Reply:
> Maybe we can just use the existing functions `WithXxx()` to set the default paris?

I don't think so.

They have different semantic meanings.

- `WithStorageClass` set current operation's storage class
- `WithDefaultStorageClass` set current storage's default storage class.

So `WithDefaultStorageClass` is not a valid pair for `Write` operation, only valid for `NewStorage` operation. And `WithStorageClass` is valid for `Write` but not for `NewStorage`.

NOTES: there is no direct conversion between `WithDefaultStorageClass` and `WithStorageClass`. `WithDefaultStorageClass` configure the `DefaultPairs` and `DefaultPairs` will be expand in corresponding operations.

If `loose_pair` is enabled, they will be ignored, otherwise, we will raise not support pairs if used incorrectly.

And to the next question: How about using `WithDefaultPairs` and `WithDefaultStorageClass` at the same time?

We will pick `DefaultPairs` first, and append pairs that parsed from `DefaultStorageClass`.

Reply:
Do you think it is necessary to explain the rationale why we need to add new pairs? Since the RFC title is about connection string.

Reply:
Nice idea.